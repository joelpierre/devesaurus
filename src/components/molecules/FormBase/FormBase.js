import React from 'react';
import PropTypes from 'prop-types';

import classNames from 'classnames';

import styles from './FormBase.module.scss';
import InputField from '../../atoms/InputField/InputField';
import SelectField from '../../atoms/SelectField/SelectField';
import FileUpload from '../../atoms/FileUpload/FileUpload';
import Row from '../../core/Row/Row';
import Flex from '../../core/Flex/Flex';
import FormField from '../FormField/FormField';
import Button from '../Buttons/Button';

const mockFormBase = [
  {
    fieldSet: {
      legend: 'FormBase Legend',
      description: 'some text here',
      fields: [
        {
          type: 'input',
          columns: {
            colMd: 6,
          },
          input: {
            label: 'label-1',
            name: 'test-1',
            type: 'text',
            placeholder: 'placeholder-1',
          },
        },
        {
          type: 'select',
          columns: {
            colMd: 6,
          },
          input: {
            label: 'label-2',
            name: 'test-2',
            placeholder: 'placeholder-1',
            options: [
              {
                name: 'option-name',
                value: 'option-value',
              },
            ],
          },
        },
        {
          type: 'file',
          columns: {},
          input: {
            name: 'file-name',
            placeholder: 'file-placeholder',
          },
        },
      ],
    },
  },
];

const FormBase = ({ formGroups }) => {
  return (
    <form
      data-test="component-form-base"
      className={classNames(styles['form-base'])}
    >
      {formGroups.map(
        ({ fieldSet: { legend, description, fields } }, index) => (
          <div key={index} className={styles['form-base__group']}>
            {legend && (
              <legend className={styles['form-base__legend']}>{legend}</legend>
            )}
            {description && (
              <p className={styles['form-base__description']}>{description}</p>
            )}
            <Row classes={styles['form-base__row']}>
              {fields.map(
                (
                  {
                    type,
                    columns: { col, colXs, colSm, colMd, colLg, colXl },
                    input: {
                      name,
                      type: inputType,
                      placeholder,
                      options,
                      ...field
                    },
                  },
                  index
                ) => (
                  <Flex
                    key={index}
                    colXl={colXl}
                    colLg={colLg}
                    colMd={colMd}
                    colSm={colSm}
                    colXs={colXs}
                    col={col || 12}
                  >
                    <FormField
                      classes={classNames(styles['form-base__form-field'])}
                    >
                      {type === 'input' && (
                        <InputField
                          name={name}
                          type={inputType}
                          placeholder={placeholder}
                          classes={classNames(styles['form-base__input-field'])}
                          {...field}
                        />
                      )}
                      {type === 'select' && (
                        <SelectField
                          name={name}
                          placeholder={placeholder}
                          classes={classNames(styles['form-base__select'])}
                          options={options}
                          {...field}
                        />
                      )}
                      {type === 'file' && (
                        <FileUpload
                          name={name}
                          placeholder={placeholder}
                          classes={classNames(styles['form-base__file-upload'])}
                          {...field}
                        />
                      )}
                    </FormField>
                  </Flex>
                )
              )}
            </Row>
          </div>
        )
      )}

      <FormField classes={classNames(styles['form-base__submit-field'])}>
        <Button behavior="action" type="submit">
          Submit
        </Button>
      </FormField>
    </form>
  );
};

FormBase.defaultProps = {
  formGroups: mockFormBase,
};

FormBase.propTypes = {
  formGroups: PropTypes.arrayOf(
    PropTypes.shape({
      fieldSet: PropTypes.shape({
        legend: PropTypes.string,
        description: PropTypes.string,
        fields: PropTypes.arrayOf(
          PropTypes.shape({
            type: PropTypes.oneOf(['input', 'select', 'file']),
            columns: PropTypes.shape({
              col: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
              colXs: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
              colSm: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
              colMd: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
              colLg: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
              colXl: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
            }),
            input: PropTypes.shape({
              name: PropTypes.string.isRequired,
              label: PropTypes.string,
              type: PropTypes.oneOf([
                'text',
                'password',
                'date',
                'email',
                'select',
                'file',
                'number',
              ]),
              placeholder: PropTypes.string,
              disabled: PropTypes.bool,
              sm: PropTypes.bool,
              md: PropTypes.bool,
              lg: PropTypes.bool,
              min: PropTypes.string,
              max: PropTypes.number,
              // Select options
              options: PropTypes.arrayOf(
                PropTypes.shape({
                  name: PropTypes.string.isRequired,
                  value: PropTypes.string.isRequired,
                })
              ),
            }),
          })
        ),
      }),
    })
  ),
};

export default FormBase;
