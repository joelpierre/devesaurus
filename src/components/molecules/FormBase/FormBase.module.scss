@import "~sass/core";

.form-base {
  $root: &;
  padding: 0;

  &__legend {
    font-weight: bold;
  }

  &__description {
    margin-bottom: 2rem !important;
  }
}
