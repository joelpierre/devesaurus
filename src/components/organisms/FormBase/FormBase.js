import React from 'react';
import PropTypes from 'prop-types';

const FormBase = (props) => {
  return (<></>);
};

FormBase.defaultProps = {};

FormBase.propTypes = {};

export default FormBase;
