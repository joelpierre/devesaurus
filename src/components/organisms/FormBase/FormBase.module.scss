@import "~sass/core";

.form-base {
  $root: &;

  width: 100%;
  position: relative;
}
