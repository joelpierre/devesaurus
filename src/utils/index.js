export const updateObject = (oldObject, updatedValues) => ({
  ...oldObject,
  ...updatedValues,
});

export const mapOverACFComponents = components => {
  components.map(component => {
    component.acf_fc_layout = component.acf_fc_layout.replace(
      'WordPressAcf_',
      ''
    );
  });
  return components;
};

export const sortWordObj = word => {
  if (word) {
    delete word.author_id;
    delete word.author;
    delete word.author_nicename;
    delete word.content;
    delete word.date_modified;
    delete word.excerpt;
    delete word.yoast;

    return word;
  }

  return false;
};

export const randTheme = () => {
  const theme = ['brand', 'alpha', 'beta', 'gamma', 'psi', 'omega'];
  const rand = randNumBetween(0, theme.length);
  return theme[rand];
};

export const randNumBetween = (min = 0, max = 10) => {
  return Math.floor(Math.random() * max) + min;
};

export const mapTaxonomyTheme = (slug = 'default') => {
  let theme;

  switch (slug) {
    case 'development':
    case 'web-development':
      theme = 'alpha';
      break;
    case 'graphic-design':
    case 'design':
      theme = 'beta';
      break;
    case 'css':
    case 'css3':
      theme = 'gamma';
      break;
    case 'html':
    case 'html5':
      theme = 'psi';
      break;
    case 'news':
      theme = 'omega';
      break;
    default:
      theme = 'brand';
  }

  return theme;
};

export const breakpoint = {
  is(s) {
    const size = s.trim();
    const sizes = {
      xsmall: '599px',
      small: '600px',
      medium: '900px',
      large: '1200px',
      xlarge: '1800px',
    };

    // console.log(window.matchMedia);

    // Eslint doesn't like you accessing hasOwnProperty directly on object.
    // https://eslint.org/docs/rules/no-prototype-builtins
    if (Object.prototype.hasOwnProperty.call(sizes, size)) {
      return window.matchMedia(`only screen and (min-width: ${sizes[size]})`)
        .matches;
    }

    throw new ReferenceError(
      `The size ${size} is not a valid breakpoint: ${JSON.stringify(sizes)}`
    );
  },
};
