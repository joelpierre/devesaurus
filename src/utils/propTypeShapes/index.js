import PropTypes from 'prop-types';

export const wordPropTypeShape = {
  word: PropTypes.string.isRequired,
  origin: PropTypes.string.isRequired,
  syllables: PropTypes.shape({
    count: PropTypes.number.isRequired,
    list: PropTypes.instanceOf(Array).isRequired,
  }).isRequired,
  pronunciation: PropTypes.string,
  results: PropTypes.arrayOf(PropTypes.shape({
    definition: PropTypes.string.isRequired,
    partOfSpeech: PropTypes.oneOf(['noun', 'verb', 'adjective', 'pronoun', 'adverb', 'preposition', 'conjunction', 'interjection']),
  })).isRequired,
  category: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    slug: PropTypes.string.isRequired,
  }),
  tags: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    slug: PropTypes.string.isRequired,
  })),
};
