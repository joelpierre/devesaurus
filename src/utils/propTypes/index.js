import PropTypes from 'prop-types';

export const wordPropTypeShape = {
  title: PropTypes.string.isRequired,
  slug: PropTypes.string.isRequired,
  terms: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      slug: PropTypes.string.isRequired,
      taxonomy: PropTypes.string.isRequired,
    })
  ),
  acf: PropTypes.shape({
    definition: PropTypes.string,
    origin: PropTypes.shape({
      label: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
    }),
    syllables: PropTypes.shape({
      count: PropTypes.string.isRequired,
      list: PropTypes.arrayOf(
        PropTypes.shape({
          item: PropTypes.string,
        })
      ).isRequired,
    }).isRequired,
    pronunciation: PropTypes.string,
  }),
};

export const columnShape = PropTypes.oneOf([
  '1',
  1,
  '2',
  2,
  '3',
  3,
  '4',
  4,
  '5',
  5,
  '6',
  6,
  '7',
  7,
  '8',
  8,
  '9',
  9,
  '10',
  10,
  '11',
  11,
  '12',
  12,
]);

export const defaultSize = (size = 'md') => ({
  size,
});

export const sizePropType = {
  size: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl', 'full']),
};

export const defaultTheme = (theme = 'brand') => ({
  theme,
});

export const themePropType = {
  theme: PropTypes.oneOf([
    'brand',
    'alpha',
    'beta',
    'gamma',
    'psi',
    'omega',
    'tint-alpha',
    'tint-beta',
    'tint-gamma',
    'tint-psi',
    'tint-omega',
    'gradient-brand',
  ]),
};

export const defaultPageTheme = (theme = 'brand') => ({
  pageTheme: theme,
});

export const pageThemePropType = {
  pageTheme: PropTypes.oneOf([
    'brand',
    'alpha',
    'beta',
    'gamma',
    'psi',
    'omega',
    'tint-alpha',
    'tint-beta',
    'tint-gamma',
    'tint-psi',
    'tint-omega',
  ]),
};
