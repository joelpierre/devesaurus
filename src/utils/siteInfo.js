import React from 'react';
import { graphql, StaticQuery } from 'gatsby';

const SiteInfo = () => (
  <StaticQuery
    query={graphql`
      {
        allWordpressSiteMetadata {
          edges {
            node {
              name
              description
            }
          }
        }
      }
    `}
    render={({ allWordpressSiteMetadata }) => (
      <>
        {allWordpressSiteMetadata.edges[0].node.name}
        <span>{allWordpressSiteMetadata.edges[0].node.description}</span>
      </>
    )}
  />
);

export default SiteInfo;
