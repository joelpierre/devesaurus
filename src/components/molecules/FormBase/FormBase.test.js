import React from 'react';
import { shallow } from 'enzyme';

import { checkProps, findByTestAttr, matchSnapshot } from '../../../utils/test';

import FormBase from './FormBase';

const defaultProps = {};

/**
 * Factory function to create a ShallowWrapper for the FormBase component.
 * @param {object} props - Component props specific to setup
 * @returns {ShallowWrapper}
 */
const setup = (props = {}) => {
  const setupProps = { ...defaultProps, ...props };
  return shallow(<FormBase {...setupProps} />);
};

describe('<FormBase/>', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = setup();
  });

  it('renders the FormBase Component without errors', () => {
    const component = findByTestAttr(wrapper, 'component-form-base');
    expect(component.length).toBe(1);
  });

  it('should render with correct props', () => {
    checkProps(FormBase, defaultProps);
  });

  it('should match snapshot', () => {
    matchSnapshot(wrapper);
  });
});
