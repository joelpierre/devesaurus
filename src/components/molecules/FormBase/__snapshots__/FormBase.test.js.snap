// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormBase/> should match snapshot 1`] = `
<form
  className="form-base"
  data-test="component-form-base"
  encType="multipart/form-data"
  id="test-search-form"
  onSubmit={[Function]}
>
  <div
    className="form-base__group"
    key="0"
  >
    <legend
      className="form-base__legend"
    >
      FormBase Legend
    </legend>
    <p
      className="form-base__description"
    >
      some text here
    </p>
    <Row
      className="form-base__row"
      column={false}
      reverse={false}
      row={true}
    >
      <Flex
        auto={false}
        col={12}
        key="0"
        lg={false}
        md={false}
        sm={false}
        type="div"
        xl={false}
        xs={false}
      >
        <FormField
          className="form-base__form-field"
        >
          <InputField
            className="form-base__input-field"
            disabled={false}
            label="label-1"
            lg={false}
            max={null}
            md={false}
            min={null}
            name="test-1"
            placeholder="placeholder-1"
            sm={false}
            type="text"
          />
        </FormField>
      </Flex>
      <Flex
        auto={false}
        col={12}
        key="1"
        lg={false}
        md={false}
        sm={false}
        type="div"
        xl={false}
        xs={false}
      >
        <FormField
          className="form-base__form-field"
        >
          <SelectField
            className="form-base__select-field"
            label="label-2"
            name="test-2"
            options={
              Array [
                Object {
                  "name": "option-name",
                  "value": "option-value",
                },
              ]
            }
            placeholder="placeholder-1"
          />
        </FormField>
      </Flex>
      <Flex
        auto={false}
        col={12}
        key="2"
        lg={false}
        md={false}
        sm={false}
        type="div"
        xl={false}
        xs={false}
      >
        <FormField
          className="form-base__form-field"
        >
          <FileUpload
            className="form-base__file-upload"
            name="file-name"
            placeholder="file-placeholder"
          />
        </FormField>
      </Flex>
    </Row>
  </div>
  <Row
    className="form-base__row"
    column={false}
    reverse={false}
    row={true}
  >
    <Flex
      auto={true}
      key="0"
      lg={false}
      md={false}
      sm={false}
      type="div"
      xl={false}
      xs={false}
    >
      <FormField
        className="form-base__buttons"
      >
        <Button
          behavior="action"
          caps={false}
          className="form-base__button"
          size="md"
          theme="brand"
          type="submit"
        >
          Submit Form
        </Button>
      </FormField>
    </Flex>
    <Flex
      auto={true}
      key="1"
      lg={false}
      md={false}
      sm={false}
      type="div"
      xl={false}
      xs={false}
    >
      <FormField
        className="form-base__buttons"
      >
        <Button
          behavior="action"
          caps={false}
          className="form-base__button"
          size="md"
          theme="beta"
          type="reset"
        >
          Reset Form
        </Button>
      </FormField>
    </Flex>
  </Row>
</form>
`;
